### PR TITLE
JDK-8277056: Combining several C2 Print* flags asserts in xmlStream::pop_tag

### PR DESCRIPTION
**Pull request by tobiasholenstein**
Sometimes the writing to xmlStream is mixed from several threads, and therefore the `xmlStream` tag stack can end up in a bad state. When this occurs, the VM crashes in `xmlStream::pop_tag` with `assert(false) failed: bad tag in log`.

The logging between `xtty->head` and `xtty->tail` is guarded by a ttyLocker which also locks `VMThread::should_terminate()` (ttyLocker is needed here to serializes the output info about the termination of the VMThread). 

The problem is that `print_metadata` and `dump_asm` may block for safepoint which then releases the ttyLocker (`break_tty_lock_for_safepoint`). When the `print_metadata` or `dump_asm` continues after the safepoint other threads could have taken the ttyLocker and may be busy printing and interfere with the tag stack of `xmlStream`.

The solution is to call `print_metadata` and `dump_asm` first and let them write to a local stringStream. Then acquire the ttyLocker to do the printing (using the local stringStream) and use a separate ttyLocker for `VMThread::should_terminate()`

The same issue was already targeted in https://bugs.openjdk.java.net/browse/JDK-8153527 but the fix was incomplete.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277056](https://bugs.openjdk.java.net/browse/JDK-8277056): Combining several C2 Print* flags asserts in xmlStream::pop_tag

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to 8140114020500d990c41c7996847ab49940b3367
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Xin Liu](https://openjdk.java.net/census#xliu) (@navyxliu - Committer)
 * [Dean Long](https://openjdk.java.net/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8203/head:pull/8203` \
`$ git checkout pull/8203`

Update a local copy of the PR: \
`$ git checkout pull/8203` \
`$ git pull https://git.openjdk.java.net/jdk pull/8203/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8203`

View PR using the GUI difftool: \
`$ git pr show -t 8203`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8203.diff">https://git.openjdk.java.net/jdk/pull/8203.diff</a>

</details>

### CODE REVIEW

**Review by vnkozlov — APPROVED** [review]
Good.

**Comment by dean-long** [comment]
Is the compiler thread running in VM mode or native mode at this point?  If it's in native mode, couldn't the tty lock get broken at any time?  It seems like we should collect the output in stringStreams, enter VM mode, lock tty, then perform output.

**Comment by tobiasholenstein** [comment]
> Is the compiler thread running in VM mode or native mode at this point? If it's in native mode, couldn't the tty lock get broken at any time? It seems like we should collect the output in stringStreams, enter VM mode, lock tty, then perform output.

The compiler thread is running in native mode. Can you elaborate why tty lock can get broken at any time in native mode?

**Comment by dean-long** [comment]
> The compiler thread is running in native mode. Can you elaborate why tty lock can get broken at any time in native mode?

GC doesn't wait for threads in native mode.  They are already considered "safe", and I was thinking break_tty_lock_for_safepoint was called by the GC thread.  But looking at the code, it looks like that isn't true.  It really is the compiler thread blocking on a safepoint.  So is it really necessary for these functions to block for a safepoint?

**Comment by TobiHartmann** [comment]
> So is it really necessary for these functions to block for a safepoint?

Isn't it just convention that a thread in native mode checks for a safepoint request when transitioning to VM mode? I.e. in this case via: `ciMetadata::print_metadata` -> `GUARDED_VM_ENTRY` -> `VM_ENTRY_MARK` -> `ThreadInVMfromNative` -> `transition_from_native` -> `SafepointMechanism::process_if_requested_with_exit_check`.

**Comment by navyxliu** [comment]
> `process_if_requested_with_exit_check`

I think this is exactly what happened in the problem described in the first place. 
... -> SafepointMechanism::process_if_requested_with_exit_check -> SafepointSynchronize::block ->   ttyLocker::break_tty_lock_for_safepoint

ttylocker may quietly get released inside of ciMetadata::print_metadata. it's not thread-safe until you regain ttylock.

File: src/hotspot/share/opto/output.cpp [hunk 1]
```diff
@@ -1871,19 +1884,15 @@ void PhaseOutput::fill_buffer(CodeBuffer* cb, uint* blk_starts) {
       }
       if (C->method() != NULL) {
         tty->print_cr("----------------------- MetaData before Compile_id = %d ------------------------", C->compile_id());
-        C->method()->print_metadata();
+        tty->print("%s", method_metadata_str.as_string());
```
Comment (navyxliu) [inline src/hotspot/share/opto/output.cpp:1887]: is print_raw() better than  print here?  at least you don't need to handle vararg.

File: src/hotspot/share/opto/output.cpp [hunk 2]
```diff
@@ -1862,6 +1862,19 @@ void PhaseOutput::fill_buffer(CodeBuffer* cb, uint* blk_starts) {
...
+      // print_metadata and dump_asm may safepoint which makes us loose the ttylock.
+      // We call them first and write to a stringStream, then we retake the lock to
+      // make sure the end tag is coherent, and that xmlStream->pop_tag is done thread safe.
+      ResourceMark rm;
+      stringStream method_metadata_str;
+      if (C->method() != NULL) {
+        C->method()->print_metadata(&method_metadata_str);
+      }
+      stringStream dump_asm_str;
+      dump_asm_on(&dump_asm_str, node_offsets, node_offset_limit);
```
Comment (navyxliu) [inline src/hotspot/share/opto/output.cpp:1874]: Does dump_asm_on also require InVM like print_metadata?  If it's not, is it easier to resume ttylock right after print_metadata?

**Comment by dean-long** [comment]
If the problem is the native --> VM transition while holding the tty lock, then how about if we enter VM mode first, then grab the tty lock?

**Comment by TobiHartmann** [comment]
> I think this is exactly what happened in the problem described in the first place.

Yes, at least that's the failure mode that @tobiasholenstein was observing.

> If the problem is the native --> VM transition while holding the tty lock, then how about if we enter VM mode first, then grab the tty lock?

I think that would work but only if there are no other safepoint checks hiding somewhere in these methods.

File: src/hotspot/share/opto/output.cpp (hunk 1 above)
Comment (tobiasholenstein) [reply src/hotspot/share/opto/output.cpp]: Agree. I changed it

File: src/hotspot/share/opto/output.cpp (hunk 2 above)
Comment (tobiasholenstein) [reply src/hotspot/share/opto/output.cpp:1874]: There is a `GUARDED_VM_ENTRY` that may safepoint in `dump_asm_on` -> `dump_on` -> `dump2` -> `print_name_on` -> `print_symbol_on`.

**Comment by tobiasholenstein** [comment]
> If the problem is the native --> VM transition while holding the tty lock, then how about if we enter VM mode first, then grab the tty lock?

Unfortunately, there are already `VM_ENTRY_MARK`'s for example in `dump_asm_on` -> `CallStaticJavaDirectNode::format` -> `CallStaticJavaDirectNode::format` -> `JVMState::print_method_with_lineno` -> `ciMethod:line_number_from_bci`. The `VM_ENTRY_MARK` requires the thread to be in native mode for the transition and cannot be called if it was already called before.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by navyxliu — APPROVED** [review]
LGTM.  I am not a reviewer.

**Comment by dean-long** [comment]
> > If the problem is the native --> VM transition while holding the tty lock, then how about if we enter VM mode first, then grab the tty lock?
> 
> Unfortunately, there are already `VM_ENTRY_MARK`'s for example in `dump_asm_on` -> `CallStaticJavaDirectNode::format` -> `CallStaticJavaDirectNode::format` -> `JVMState::print_method_with_lineno` -> `ciMethod:line_number_from_bci`. The `VM_ENTRY_MARK` requires the thread to be in native mode for the transition and cannot be called if it was already called before.

There are ways to fix that.  I see other methods that use GUARDED_VM_ENTRY, for example.  But maybe it's not worth it.

**Review by dean-long — APPROVED** [review]

**Comment by tobiasholenstein** [comment]
@dean-long , @TobiHartmann , @vnkozlov and @navyxliu thanks for the reviews!

**Pull request closed without merging**